pkg/reconcilermanager/controllers/rootsync_controller.go

```diff
@@ -492,6 +495,90 @@ func (r *RootSyncReconciler) mapMembershipToRootSyncs() func(client.Object) []re
 	}
 }
 
+// mapConfigMapsToRootSyncs handles updates to referenced ConfigMaps
+func (r *RootSyncReconciler) mapConfigMapsToRootSyncs(obj client.Object) []reconcile.Request {
+	objRef := client.ObjectKeyFromObject(obj)
+
+	// Ignore changes from other namespaces.
+	if objRef.Namespace != configsync.ControllerNamespace {
+		return nil
+	}
+
+	ctx := context.Background()
+	allRootSyncs := &v1beta1.RootSyncList{}
+	if err := r.client.List(ctx, allRootSyncs); err != nil {
+		klog.Error("failed to list all RootSyncs for %s (%s): %v",
+			obj.GetObjectKind().GroupVersionKind().Kind, objRef, err)
+		return nil
+	}
+
+	var result []reconcile.Request
+	for _, rs := range allRootSyncs.Items {
+		if rs.Spec.SourceType != string(v1beta1.HelmSource) {
+			// we are only watching ConfigMaps to retrigger helm-sync, so we can ignore
+			// other source types
+			continue
+		}
+		if rs.Spec.Helm == nil || len(rs.Spec.Helm.ValuesFrom) == 0 {
+			continue
+		}
+
+		var referenced bool
+		var key string
+		for _, vf := range rs.Spec.Helm.ValuesFrom {
+			if vf.Name == objRef.Name {
+				referenced = true
+				key = vf.Key
+				break
+			}
+		}
+
+		if !referenced {
+			continue
+		}
+
+		if err := validate.CheckConfigMapKeyExists(ctx, r.client, objRef, key, &rs); err != nil {
+			// If the ConfigMap was deleted or it no longer has the specified data key, we skip restarting the pod
+			// to avoid losing or modifying already synced resources
+			klog.Errorf("ConfigMap ref validation error: %s", err.Error())
+			// requeue the reconcile to rerun validation checks and status setting
+			result = append(result, requeueRootSyncRequest(obj, &rs)...)
+
+		}
+
+		rsRef := client.ObjectKeyFromObject(&rs)
+		reconcilerRef := types.NamespacedName{
+			Namespace: configmanagement.ControllerNamespace,
+			Name:      core.RootReconcilerName(rsRef.Name),
+		}
+
+		var reconciler appsv1.Deployment
+		cl := r.reconcilerBase.client
+		if err := cl.Get(ctx, reconcilerRef, &reconciler); err != nil {
+			klog.Errorf("failed to get reconciler reference: %s", err.Error())
+			if apierrors.IsNotFound(err) {
+				// if the root-reconciler doesn't exist, an updated ConfigMap should trigger a reconcile
+				// to create one
+				result = append(result, requeueRootSyncRequest(obj, &rs)...)
+			}
+			continue
+		}
+
+		// A rereconcile doesn't necessarily update an existing root-reconciler deployment, so
+		// the root-reconciler pods need to be explicitly restarted to pick up the ConfigMap update.
+		reconciler.Spec.Template.ObjectMeta.Annotations["kubectl.kubernetes.io/restartedAt"] = time.Now().Format(time.RFC3339)
```

Comment:
What the heck is this?

Is this trying to evict the pods or just restart them?

Reply:
It just restarts them, otherwise the pod doesn't pick up the change in the ConfigMap. Inspiration came from this: https://stackoverflow.com/questions/61335318/how-to-restart-a-deployment-in-kubernetes-using-go-client, as I was looking for the equivalent of `kubectl rollout restart`.

Reply:
We shouldn't need this given that mounted ConfigMap is updated automatically https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/#mounted-configmaps-are-updated-automatically There might be some delay before it detects the change.

Reply:
Discussed offline, another reason to restart the pods is that currently helm-sync only compares version to trigger refresh and rerender. If we don't restart the pod, we will have to compare all the valuesFiles for changes as well.

We will recommend to the best practice of treating ConfigMaps as immutable, and leave the restart as a workaround for now.

Reply:
One of the reasons we stopped using ConfigMaps to configure the reconciler was that changing the ConfigMap doesn’t automatically update the Pods. But worse, it also doesn’t update the Deployment, which means the generation doesn’t change. So there’s no way to know if the Deployment status reflects the latest ConfigMap change. 

This will probably be a problem for customers who want to wait until their config change takes effect. 

Do we have a way to communicate that reconciliation?

Reply:
Could we delete the Pod instead to force a restart?